### ISSUE TITLE
Issues with deleting sections/rows in ASTableView

### ISSUE THREAD

**Issue by tomizimobile** [facebook/AsyncDisplayKit#869]
I've created a simple test project that shows the issue: https://github.com/tomizimobile/ASDKSearchIssue

The project is pretty basic - just an `ASTableView` that shows some randomly generated names, and a search bar to filter the results. Using UIKit, everything works as expected, but using AsyncDisplayKit, a crash occurs when trying to delete sections/rows (using `tableView.beginUpdates()` + update calls) or reload the data (using `tableView.reloadData()`) after the data source has been filtered. It appears to be calling `- tableView:titleForHeaderInSection:` with a section that is out of range for the filtered data.

**Comment by tettoffensive** [comment]
I am also having trouble with something similar. `- (ASCellNode *)tableView:(ASTableView *)tableView nodeForRowAtIndexPath:(NSIndexPath *)indexPath` is being called at some point when the data source has changed and indexPath.row is out of index

**Comment by Adlai-Holler** [comment]
@tomizimobile Thanks for posting this issue, and sweet test project!

I'm investigating it currently. If I type "T", the app requests to delete sections {0, 5, 11} but we end up only deleting section 11 and thus requesting the title for now-gone section 10. The other two indices are dropped somewhere in `+[_ASHierarchyChangeSet sortAndCoalesceChanges:]`.

**Comment by Adlai-Holler** [comment]
@tomizimobile @tettoffensive As the author of the broken code, I'm really sorry about this issue. I believe #884 will resolve it – it fixed the sample project on my machine. If you have time, could you try replacing the line in your Podfile with

`pod 'AsyncDisplayKit', :git => 'https://github.com/Adlai-Holler/AsyncDisplayKit', :branch => 'FixTableViewUpdatingIssue'`

and running `pod update AsyncDisplayKit` to see if the proposed PR fixes the issue?

**Comment by Adlai-Holler** [comment]
Sorry, put a quick hold on that. There's another potential issue that I'd like to fix at the same time. Shouldn't be more than an hour.

**Comment by tomizimobile** [comment]
@Adlai-Holler That fix seems to be handling deletions pretty fine, but I noticed when putting it into my test project, it has weird behavior when inserting sections. I searched for "T" to filter the results, which filters data, and removes sections "A", "J", and "Y". Now, when deleting the search text, all the data is back but the rows in sections "A", "J", and "Y" are doubled-up.

**Comment by Adlai-Holler** [comment]
OK thanks for saying so. It looks like a separate issue but I'm attempting to fix it right now. I'll post again soon with an update.

**Comment by Adlai-Holler** [comment]
NOTE: The issue causing the doubling appears to be this:

If you insert a section during a batch update and also insert rows into that section, we will insert those rows twice. It's an open question which animation options UIKit uses in that case – the options given for the section insert or for the row insert – I'll assume it uses the options for the section insert.

**Comment by Adlai-Holler** [comment]
@tomizimobile OK could you give the latest on that branch a try? I've fixed the duplication issue as well.

**Comment by tomizimobile** [comment]
Looks great! I only inserted the rows along with the section because I seem to remember needing to do that in an older version of ASDK... it might be unnecessary but your branch handles it just fine. Thanks for taking a look at this.

**Comment by appleguy** [comment]
@tomizimobile thanks so much for not only providing the project, but helping with followup testing of the patch - it really helps us move quickly and more confidently.  Bummed to hear you needed a weird workaround for a prior version, but sadly I totally understand as a filtering typeahead is a sure way to bust the shady implementation of this code from 1.2.x and even some of the more recent versions :-P.

Of course, it turns out that workaround allowed us to make this code a lot better — I am very glad to hear that we caught the issue of editing a section that hasn't yet been added, since that should absolutely work reliably even if it's possible to avoid in your app.

Thanks as well @tettoffensive for adding more details!

Wondering, have you encountered any other problems with the framework or have feature requests?

**Comment by tomizimobile** [comment]
@appleguy Happy to provide the sample project and follow up with testing - I know how much easier it makes debugging. 

I've only ran across a couple previously reported issues (I've reported https://github.com/facebook/AsyncDisplayKit/issues/860, and https://github.com/facebook/AsyncDisplayKit/issues/261 is a longstanding issue). I might have another issue with an occasional deadlock on collection view updates, but I haven't reliably tracked it down, and I'm not sure if it's an issue with my own code.

Really loving the recent layout spec additions. I don't have any feature requests at the moment, but I'll post them here if I do in the future.
